### ISSUE TITLE
Use DI to Configure IExternalScopeProvider

### ISSUE THREAD

**Issue by TheXenocide** [dotnet/runtime#36090]
### Is your feature request related to a problem? Please describe.
I am trying to collate ambient and global diagnostic data to be provided as structured event data with logs and have found that, despite the fact that an `IExternalScopeProvider` interface exists, there's no way to configure the default `LoggerFactory` to use anything but the default `LoggerExternalScopeProvider`. I literally only need to supply one extra state object to loggers that support external scope which, ideally, would involve inheriting the default provider and overriding `ForEachScope` (or some similar such facility that would let me tie in to all scope-aware loggers to provide one additional global-but-contextual set of metadata).

### Describe the solution you'd like
The default `LoggerExternalScopeProvider` should be registered using the same `TryAdd`/`Singleton` behavior seen in `LoggingServiceCollectionExtensions.AddLogging` as is used for other default logging interface implementations and `LoggerFactory` should receive it as an injected constructor parameter.

### Describe alternatives you've considered
In the meantime (unless there's a more advisable solution) I think I'm going to wind up creating a decorator (since none of the methods I need to override are virtual) implementation of `ILoggerFactory` whose only purpose is to wrap the implementation of `LoggerFactory` so that, after calling the default implementation's `AddProvider` method I can call `SetScopeProvider` to replace the one the default implementation used with a decorator implementation of `IExternalScopeProvider` in which the implementation of `ForEachScope` will call the callback with a state object gathered from our custom contextual diagnostic scope and global values either before or after calling the default implementation.

### Additional context
From the archived repository: https://github.com/aspnet/Logging/issues/875
Related: https://github.com/aspnet/Extensions/issues/644
Question that prompted my deep dive: https://github.com/aspnet/Extensions/issues/2398
PR containing an alternative non-DI solution: https://github.com/aspnet/Extensions/pull/2170

**Comment by TheXenocide** [comment]
@dferretti RE:
> but the main thing I'd like to be able to do is to inject this IExternalScopeProvider into my other services (middlewares of sorts) and know that any logger that supports external scope will make use of whatever scopes they add.

I believe to accomplish that you can just use ILogger.BeginScope in your middleware.

**Comment by dferretti** [comment]
You're right! I had forgotten what I originally was looking for when this issue notified me the other day. The logger BeginScope works fine for all that. I remember now what I was looking for at the time - a way to inject my own IExternalScopeProvider that would be used, so I could also inject it to other services for outgoing events/messages. Basically an async context might accumulate several scoped logging values as it progresses, and for example when it got to a part where we were going to send something over rabbitmq to another service we could include these values and recreate them in the downstream service. But after a while we settled on a different approach that didn't depend on pulling scoped log context values, and instead was more similar to https://github.com/dotnet/aspnetcore/pull/7921

**Comment by maryamariyan** [comment]
Related to https://github.com/dotnet/runtime/issues/36546

**Comment by rosebyte** [comment]
@KalleOlaviNiemitalo is right, the core ask here was implemented by #67520 and shipped in .NET 7.

The one part of the original request that wasn't implemented is registering the default provider in DI via `TryAddSingleton`. In practice that's a single line:

```csharp
services.AddSingleton<IExternalScopeProvider, LoggerExternalScopeProvider>();
```

Making that the default isn't free, though: `LoggerFactory` only creates a scope provider when a registered provider implements `ISupportExternalScope`. Registering one unconditionally would make `_scopeProvider` non-null in every app, adding a scope logger where there currently isn't one. That's a behavioural change for existing users.
